floorp: drop, replace with floorp-bin; init at 12.1.4

christoph-heiss opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Fixes #438494.
Fixes #431292.

It's a -bin variant of Floorp, which previously we built from source. Upstream became pretty much unfeasible to build from source starting with the 12.x release line, so let's replace it with a `-bin` variant.

Long-term this makes it also a lot easier on maintenance and Hydra, both resource and time-wise.

I have no macOS device(s), so would be heavily appreciated if someone could test that.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [x] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

christoph-heiss committed:
floorp-bin: init at 12.1.4

Fixes #438494.
Fixes #431292.

It's a -bin variant of floorp, which previously we built from source.
Upstream became pretty much unfeasible to build from source starting
with the 12.x release line, so let's replace it with a `-bin` variant.

Long-term this makes it also a lot easier on maintenance and Hydra,
both resource and time-wise.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

qweered reviewed (approved):
Approved automatically following the successful run of `nixpkgs-review`.

dtomvan commented on pkgs/by-name/fl/floorp-bin-unwrapped/update.sh:
```diff
@@ -0,0 +1,61 @@
+fi
+
+hashes="$(mktemp)"
+curl --silent --location -o "$hashes" "https://github.com/$owner/$repo/releases/download/$latestTag/hashes.txt"
```
In order to get the URLs to the artifacts in combination with the hashes, I'd recommend using `jq` on the output of `curl https://api.github.com/repos/floorp-projects/floorp/releases/latest` instead of parsing their weird `hashes.txt` format. You could generate an entire `sources.json` with:
```bash
curl https://api.github.com/repos/floorp-projects/floorp/releases/latest \
    | jq '
        {
            version: .tag_name[1:],
            sources: (
                .assets 
                    | map(select(.name | (endswith(".tar.xz") or endswith(".dmg"))) 
                          | {url: .browser_download_url, sha256: (.digest | split(":").[1])})
                    | map({
                        key: .url | 
                            if contains("aarch64") then
                                "aarch64-linux"
                            elif contains("amd64") then
                                "x86_64-linux"
                            elif contains("macOS") then
                                "darwin"
                            else null end,
                        value: .
                    })
                    | from_entries
            )
        }
    ' > "$dirname/sources.json"
```

(welp, I wanted to write a small jq query but ended up rewriting all of this update script 😅 )

This yields:
```json
{
  "version": "12.1.4",
  "sources": {
    "aarch64-linux": {
      "url": "https://github.com/Floorp-Projects/Floorp/releases/download/v12.1.4/floorp-linux-aarch64.tar.xz",
      "sha256": "3221bc1299edacf061f0c66a3543532fd67ef0b7f998f4c41b4c7f062f79685b"
    },
    "x86_64-linux": {
      "url": "https://github.com/Floorp-Projects/Floorp/releases/download/v12.1.4/floorp-linux-amd64.tar.xz",
      "sha256": "86a5d86a52f3778ca3d3b7b250c3e36820ab792232e8dd8cd4482586e9c1f2f1"
    },
    "darwin": {
      "url": "https://github.com/Floorp-Projects/Floorp/releases/download/v12.1.4/floorp-macOS-universal.dmg",
      "sha256": "482723722b6bab68ce5353d49502b1db6d3d43d3f918de273ef8de77c5d7f03a"
    }
  }
}
```

dtomvan reviewed (approved):
> Upstream became pretty much unfeasible to build from source starting with the 12.x release line, so let's replace it with a `-bin` variant.

Yeah... I tried to do the same for chaotic-nyx, but also found out that deno-only projects cannot really get vendored... it's a shame. I also cannot test darwin, but since it's a binary package I think a successful build is enough to me.

I did provide some feedback, but that's far from a blocker to me. Looks good for the most part. Thanks!

misuzu commented on pkgs/by-name/fl/floorp-bin-unwrapped/package.nix:
```diff
@@ -0,0 +1,133 @@
+    if stdenv.hostPlatform.isDarwin then
+      ''
+        mkdir -p $out/Applications
+        mv Firefox*.app "$out/Applications/Floorp.app"
```
```suggestion
        mv Floorp.app "$out/Applications/Floorp.app"
```

Fix for
```
building '/nix/store/p5bznpimv1hwiqkl3pk0v3jizglw3zry-floorp-bin-unwrapped-12.1.4.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/9v2f51iy2kafh0ihbdxm4m2rgv874ay2-floorp-macOS-universal.dmg
source root is .
Running phase: patchPhase
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
no configure script, doing nothing
Running phase: buildPhase
no Makefile or custom buildPhase, doing nothing
Running phase: glibPreInstallPhase
Running phase: glibPreInstallPhase
Running phase: installPhase
mv: missing destination file operand after '/nix/store/1d7q14gbwjwsrzdba98l5ad1flinc39a-floorp-bin-unwrapped-12.1.4/Applications/Floorp.app'
Try 'mv --help' for more information.
```

The built package doesn't start though

dtomvan reviewed (approved):
Looks all good to me, works, but I have no idea what it's like on darwin. Can't test that either, sorry.

christoph-heiss committed:
floorp: drop

Will be replaced with a -bin build, which becomes necessary due to
upstream becoming unfeasible to build from source starting with the 12.x
release line.

An appropriate `throw` is also added, to advise users to upgrade to the
new -bin variant.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>
